R/epi_slide_opt_archive.R

```diff
@@ -0,0 +1,230 @@
+#'
+#' library(dplyr)
+#' grp_updates <- bind_rows(
+#'   tibble(version = 10, time_value = 1:20, value = 1:20),
+#'   tibble(version = 12, time_value = 4:5, value = 5:4),
+#'   tibble(version = 13, time_value = 8, value = 9),
+#'   tibble(version = 14, time_value = 11, value = NA),
+#'   tibble(version = 15, time_value = -10, value = -10),
+#'   tibble(version = 16, time_value = 50, value = 50)
+#' ) %>%
+#'   mutate(across(c(version, time_value), ~ as.Date("2020-01-01") - 1 + .x))
+#'
+#' f <- purrr::partial(data.table::frollmean, algo = "exact")
+#'
+#' grp_updates %>%
+#'   epiprocess:::epi_slide_opt_archive_one_epikey(
+#'     "value", f, "data.table", 2L, 0L, "day", "slide_value"
+#'   )
+#'
+#' @keywords internal
+epi_slide_opt_archive_one_epikey <- function(
+    grp_updates,
+    in_colnames,
+    f_dots_baked, f_from_package, before, after, time_type,
+    out_colnames) {
+  grp_updates_by_version <- grp_updates %>%
+    nest(.by = version, .key = "subtbl") %>%
+    arrange(version)
+  unit_step <- unit_time_delta(time_type)
+  prev_inp_snapshot <- NULL
+  prev_out_snapshot <- NULL
+  result <- map(seq_len(nrow(grp_updates_by_version)), function(version_i) {
+    version <- grp_updates_by_version$version[[version_i]]
+    inp_update <- grp_updates_by_version$subtbl[[version_i]]
+    inp_snapshot <- tbl_patch(prev_inp_snapshot, inp_update, "time_value")
+    if (before == Inf) {
+      if (after != 0) {
+        cli_abort('.window_size = Inf is only supported with .align = "right"',
+          class = "epiprocess__epi_slide_opt_archive__inf_window_invalid_align"
+        )
+      }
+      # We need to use the entire input snapshot range, filling in time gaps. We
+      # shouldn't pad the ends.
+      slide_t_min <- min(inp_snapshot$time_value)
+      slide_t_max <- max(inp_snapshot$time_value)
+    } else {
+      # If the input had updates in the range t1..t2, this could produce changes
+      # in slide outputs in the range t1-after..t2+before, and to compute those
+      # slide values, we need to look at the input snapshot from
+      # t1-after-before..t2+before+after. nolint: commented_code_linter
+      inp_update_t_min <- min(inp_update$time_value)
+      inp_update_t_max <- max(inp_update$time_value)
+      slide_t_min <- inp_update_t_min - (before + after) * unit_step
+      slide_t_max <- inp_update_t_max + (before + after) * unit_step
+    }
+    slide_nrow <- time_delta_to_n_steps(slide_t_max - slide_t_min, time_type) + 1L
+    slide_time_values <- slide_t_min + 0L:(slide_nrow - 1L) * unit_step
+    slide_inp_backrefs <- vec_match(slide_time_values, inp_snapshot$time_value)
+    # Get additional values needed from inp_snapshot + perform any NA
+    # tail-padding needed to make slider results a fixed window size rather than
+    # adaptive at tails + perform any NA gap-filling needed:
+    slide <- inp_snapshot[slide_inp_backrefs, ]
+    slide$time_value <- slide_time_values
+    if (f_from_package == "data.table") {
```

Comment:
This is the only branch that's actually tested right?

Reply:
Good catch!  I've done some ad-hoc testing with other functions but actual tests are looking pretty homogeneous...  Should probably look into code coverage tools again.